tests/MongoDB.Driver.Tests/Specifications/server-discovery-and-monitoring/ServerDiscoveryAndMonitoringIntegrationTestRunner.cs

```diff
@@ -0,0 +1,103 @@
+
+        // public methods
+        public void ConfigureFailPoint(IServer server, ICoreSessionHandle session, BsonDocument failCommand)
+        {
+            var failPoint = FailPoint.Configure(server, session, failCommand);
+            Disposables.Add(failPoint);
+        }
+
+        public async Task ConfigureFailPointAsync(IServer server, ICoreSessionHandle session, BsonDocument failCommand)
+        {
+            var failPoint = await Task.Run(() => FailPoint.Configure(server, session, failCommand)).ConfigureAwait(false);
+            Disposables.Add(failPoint);
+        }
+
+        [SkippableTheory]
+        [ClassData(typeof(TestCaseFactory))]
+        public void Run(JsonDrivenTestCase testCase)
+        {
+            SetupAndRunTest(testCase);
+        }
+
+        // protected methods
+        protected override JsonDrivenTestFactory CreateJsonFactory(IMongoClient mongoClient, string databaseName, string collectionName, Dictionary<string, object> objectMap, EventCapturer eventCapturer)
+        {
+            return new JsonDrivenTestFactory(this, mongoClient, databaseName, collectionName, bucketName: null, objectMap, eventCapturer);
+        }
+
+        protected override List<object> ExtractEventsForAsserting(EventCapturer eventCapturer)
+        {
+            return base
+                .ExtractEventsForAsserting(eventCapturer)
+                .Where(@event => @event is CommandStartedEvent) // apply events asserting only for this event
+                .ToList();
+        }
+
+        protected override EventCapturer InitializeEventCapturer(EventCapturer eventCapturer)
+        {
+            var doNotCaptureEvents = DefaultCommandsToNotCapture;
+            doNotCaptureEvents.Add("configureFailPoint");
+            doNotCaptureEvents.Add("replSetStepDown");
+            return eventCapturer.Capture<CommandStartedEvent>(e => !doNotCaptureEvents.Contains(e.CommandName))
+                .Capture<ServerDescriptionChangedEvent>()
+                .Capture<ConnectionPoolClearedEvent>();
+        }
+
+        // nested types
+        private class TestCaseFactory : JsonDrivenTestCaseFactory
+        {
+            // protected properties
+            protected override string PathPrefix => "MongoDB.Driver.Tests.Specifications.server_discovery_and_monitoring.tests.integration.";
+
+            // protected methods
+            protected override IEnumerable<JsonDrivenTestCase> CreateTestCases(BsonDocument document)
+            {
+                var testCases = base.CreateTestCases(document);
+                foreach (var testCase in testCases)
+                {
+                    foreach (var async in new[] { false, true })
+                    {
+                        var name = $"{testCase.Name}:async={async}";
+                        var test = testCase.Test.DeepClone().AsBsonDocument.Add("async", async);
```

Comment:
@DmitryLukyanov
Given the issues we've been having with running the sync tests and async tests against the same cluster, I wonder if it'd be worth it append the word "async" to the appName of any failpoint associated with an async test, in order to reduce the chance that a leftover failpoint from a sync run would affect an async run?

Reply:
>it'd be worth it append the word "async" to the appName of any failpoint associated with an async test, in order to reduce the chance that a leftover failpoint from a sync run would affect an async run?

I did it in debugging:

        private void UpdateDocuments(BsonDocument document, bool async)
        {
            if (document.TryGetElement("appname", out var appName) || document.TryGetElement("appName", out appName))
            {
                document[appName.Name] = appName.Value + $"Async{async}";
            }

            foreach (var element in document.Elements)
            {
                if (element.Value is BsonDocument doc)
                {
                    UpdateDocuments(doc, async);
                }
                else if (element.Value is BsonArray arr)
                {
                    foreach (var a in arr)
                    {
                        if (a is BsonDocument d)
                        {
                            UpdateDocuments(d, async);
                        }
                        else if (a is BsonString)
                        {
                        }
                        else
                        {
                            throw new SkipException("Skip");
                        }
                    }
                }

            }
        }

but not sure that we need to leave it in the final version, I would say more no than yes, at some moment maybe appName will be used in some other test steps and it will be not so obvious that we need to modify the above code too because the spec will not know about this change at all